File: src/main/java/br/unb/cic/bionimbuz/plugin/PluginInfo.java
/*
    BioNimbuZ is a federated cloud platform.
    Copyright (C) 2012-2015 Laboratory of Bioinformatics and Data (LaBiD), 
    Department of Computer Science, University of Brasilia, Brazil

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package br.unb.cic.bionimbuz.plugin;

import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.codehaus.jackson.map.ObjectMapper;

import br.unb.cic.bionimbuz.p2p.Host;

public class PluginInfo implements PluginOps {

    private String id;
    private String InstanceName;
    private int privateCloud;
    private Host host;
    private long uptime;
    private Double latency = 0d;
    private double costpergiga;
    private long timestamp;
    private Integer numCores;
    private Integer numNodes;
    private Integer numOccupied;
    private Double ranking = 0d;
    private Float fsSize;
    private double storagecost;
    private Float fsFreeSize;
    private Double memoryTotal;
    private Double memoryFree;
    private Double currentFrequencyCore;
    // frequency in Hz
    private Double factoryFrequencyCore;
    private List<PluginService> services;
    private double costPerHour = Double.MAX_VALUE;
    private Double bandwidth;
    private String ip;
    private String provider;

	public double gpuMemoryTotal,
			gpuMemoryMaxFrequency,
			gpuMemoryBus,
			gpuMemoryBandwith,
			gpuFloatingPointPerf;
	public double gpuMaxFrequency;


    public PluginInfo() {
    }

    public PluginInfo(String id) {
        this.id = id;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getInstanceName() {
        return InstanceName;
    }

    public void setInstanceName(String instanceName) {
        this.InstanceName = instanceName;
    }

    public Host getHost() {
        return host;
    }

    public void setHost(Host host) {
        this.host = host;
    }

    public double getCostPerGiga() {
        return costpergiga;
    }

    public void setCostPerGiga(double costpergiga) {
        this.costpergiga = costpergiga;
    }

    public long getUptime() {
        return System.currentTimeMillis() - uptime;
    }

    public void setUptime(long uptime) {
        this.uptime = uptime;

    }

    public Double getLatency() {
        return latency;
    }

    public void setLatency(Double latency) {
        this.latency = latency;

    }

    public double getStorageCost() {
        return storagecost;
    }

    public void setStorageCost(double storagecost) {
        this.storagecost = storagecost;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    public Float getFsFreeSize() {
        return fsFreeSize;
    }

    public void setFsFreeSize(Float fsFreeSize) {
        this.fsFreeSize = fsFreeSize;
    }

    public Integer getNumCores() {
        return numCores;
    }

    public void setNumCores(Integer numCores) {
        this.numCores = numCores;
    }

    public Integer getNumNodes() {
        return numNodes;
    }

    public void setNumNodes(Integer numNodes) {
        this.numNodes = numNodes;
    }

    public Float getFsSize() {
        return fsSize;
    }

    public void setFsSize(Float fsSize) {
        this.fsSize = fsSize;
    }

    public Integer getNumOccupied() {
        return numOccupied;
    }

    public void setNumOccupied(Integer numOccupied) {
        this.numOccupied = numOccupied;
    }

    public List<PluginService> getServices() {
        return services;
    }

    public void setServices(List<PluginService> services) {
        this.services = services;
    }

    public PluginService getService(String serviceId) {
        for (PluginService service : getServices()) {
            if (service.getId().equals(serviceId)) {
                return service;
            }
        }
        return null;
    }

    public Double getRanking() {
        return ranking;
    }

    public void setRanking(Double ranking) {
        this.ranking = ranking;
    }

    public Double getCurrentFrequencyCore() {
        return currentFrequencyCore;
    }

    public void setCurrentFrequencyCore(Double frequencyCore) {
        this.currentFrequencyCore = frequencyCore;
    }

    public Double getFactoryFrequencyCore() {
        return factoryFrequencyCore;
    }

    public void setFactoryFrequencyCore(Double frequencyCore) {
        this.factoryFrequencyCore = frequencyCore;
    }

    public Double getMemoryFree() {
        return memoryFree;
    }

    public void setMemoryFree(Double memoryFree) {
        this.memoryFree = memoryFree;
    }

    public Double getMemoryTotal() {
        return memoryTotal;
    }

    public void setMemoryTotal(Double memoryTotal) {
        this.memoryTotal = memoryTotal;
    }

    @Override
    public boolean equals(Object object) {
        if (this == object) {
            return true;
        }

        if (!(object instanceof PluginInfo)) {
            return false;
        }

        PluginInfo other = (PluginInfo) object;

        return this.id.equals(other.id);
    }

    public int getPrivateCloud() {
        return privateCloud;
    }

    public void setPrivateCloud(int privateCloud) {
        this.privateCloud = privateCloud;
    }

    public void setCostPerHour(double costPerHour) {
        this.costPerHour = costPerHour;
    }

    public double getCostPerHour() {
        return costPerHour;
    }

    public Double getBandwidth() {
        return bandwidth;
    }

    public void setBandwidth(Double bandwidth) {
        this.bandwidth = bandwidth;
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    //Alterado para retornar os valores que serão gravados no znode peer old: id
    @Override
    public String toString() {
        try {
            return new ObjectMapper().writeValueAsString(this);
        } catch (Exception ex) {
            Logger.getLogger(PluginInfo.class.getName()).log(Level.SEVERE, null, ex);
        }
        return null;
    }

    public String getIp() {
        return ip;
    }

    public void setIp(String ip) {
        this.ip = ip;
    }

    public String getProvider() {
        return provider;
    }

    public void setProvider(String provider) {
        this.provider = provider;
    }

    public String Serialize(){
	    String ret= "PLUGIN_INFO";
	    ret+= '\n';
	    
	    ret+= "id=";
	    ret+= id;
	    ret+= '\n';
	    
	    ret+= "instanceName=";
	    ret+= getInstanceName();
	    ret+= '\n';
	    
	    ret+= "privateCloud=";
	    ret+= String.valueOf(privateCloud);
	    ret+= '\n';
	    
	    ret+= "host=";
	    ret+= host.getAddress();
	    ret+= ':';
	    ret += String.valueOf(host.getPort());
	    ret+= '\n';
	    
	    ret+= "upTime=";
	    ret+= String.valueOf(getUptime());
	    ret+= '\n';
	    
	    ret+= "latency=";
	    ret+= String.valueOf(latency);
	    ret+= '\n';
	    
	    ret+= "costPerGiga=";
	    ret+= String.valueOf(costpergiga);
	    ret+= '\n';
	    
	    ret+= "timestamp=";
	    ret+= String.valueOf(timestamp);
	    ret+= '\n';
	    
	    ret+= "numCores=";
	    ret+= String.valueOf(numCores);
	    ret+= '\n';
	    
	    ret+= "numNodes=";
	    ret+= String.valueOf(numNodes);
	    ret+= '\n';
	    
	    ret+= "numOccupied=";
	    ret+= String.valueOf(numOccupied);
	    ret+= '\n';
	    
	    ret+= "ranking=";
	    ret+= String.valueOf(ranking);
	    ret+= '\n';
	    
	    ret+= "fsSize=";
	    ret+= String.valueOf(fsSize);
	    ret+= '\n';
	    
	    ret+= "memoryTotal=";
	    ret+= String.valueOf(memoryTotal);
	    ret+= '\n';
	    
	    ret+= "memoryFree=";
	    ret+= String.valueOf(memoryFree);
	    ret+= '\n';
	    
	    ret+= "currentFrequencyCore=";
	    ret+= String.valueOf(currentFrequencyCore);
	    ret+= '\n';
	    
	    ret+= "costPerHour=";
	    ret+= String.valueOf(costPerHour);
	    ret+= '\n';
	    
	    ret+= "bandwith=";
	    ret+= String.valueOf(getBandwidth());
	    ret+= '\n';
	    
	    ret+= "ip=";
	    ret+= ip;
	    ret+= '\n';
	    
	    ret+= "provider=";//obs:: n pode ter \n no provider
	    ret+= provider;
	    ret+= '\n';
	    
	    //adições para suporte de GPU
	    ret+= "gpuMemoryTotal=";
	    ret+= String.valueOf(gpuMemoryTotal);
	    ret+= '\n';
	    
	    ret+= "gpuMemoryMaxFrequency=";
	    ret+= String.valueOf(gpuMemoryMaxFrequency);
	    ret+= '\n';
	    
	    ret+= "gpuMemoryBus=";
	    ret+= String.valueOf(gpuMemoryBus);
	    ret+= '\n';
	    
	    ret+= "gpuMemoryBandwith=";
	    ret+= String.valueOf(gpuMemoryBandwith);
	    ret+= '\n';
	    
	    ret+= "gpuFloatingPointPerf=";
	    ret+= String.valueOf(gpuFloatingPointPerf);
	    ret+= '\n';
	    
	    ret+= "gpuMaxFrequency=";
	    ret+= String.valueOf(gpuMaxFrequency);
	    ret+= '\n';
	    
	    return ret;
    }

}

File: src/main/java/br/unb/cic/bionimbuz/services/sched/policy/impl/Cpp/src/PluginInfo.cpp
#include <stdio.h>
#include <string.h>
#include <math.h>
#include "PluginInfo.h"
#include "Error.h"

#define TEMP_BUFFER_SIZE (200)

std::string PluginInfo::Serialize(void){
	std::string ret= "PLUGIN_INFO";
	ret+= '\n';
	
	ret+= "id=";
	ret+= id;
	ret+= '\n';
	
	ret+= "instanceName=";
	ret+= instanceName;
	ret+= '\n';
	
	ret+= "privateCloud=";
	ret+= std::to_string(privateCloud);
	ret+= '\n';
	
	ret+= "host=";
	ret+= host.adderss;
	ret+= ':';
	ret += std::to_string(host.port);
	ret+= '\n';
	
	ret+= "upTime=";
	ret+= std::to_string(upTime);
	ret+= '\n';
	
	ret+= "latency=";
	ret+= std::to_string(latency);
	ret+= '\n';
	
	ret+= "costPerGiga=";
	ret+= std::to_string(costPerGiga);
	ret+= '\n';
	
	ret+= "timestamp=";
	ret+= std::to_string(timestamp);
	ret+= '\n';
	
	ret+= "numCores=";
	ret+= std::to_string(numCores);
	ret+= '\n';
	
	ret+= "numNodes=";
	ret+= std::to_string(numNodes);
	ret+= '\n';
	
	ret+= "numOccupied=";
	ret+= std::to_string(numOccupied);
	ret+= '\n';
	
	ret+= "ranking=";
	ret+= std::to_string(ranking);
	ret+= '\n';
	
	ret+= "fsSize=";
	ret+= std::to_string(fsSize);
	ret+= '\n';
	
	ret+= "memoryTotal=";
	ret+= std::to_string(memoryTotal);
	ret+= '\n';
	
	ret+= "memoryFree=";
	ret+= std::to_string(memoryFree);
	ret+= '\n';
	
	ret+= "currentFrequencyCore=";
	ret+= std::to_string(currentFrequencyCore);
	ret+= '\n';
	
	ret+= "costPerHour=";
	ret+= std::to_string(costPerHour);
	ret+= '\n';
	
	ret+= "bandwith=";
	ret+= std::to_string(bandwith);
	ret+= '\n';
	
	ret+= "ip=";
	ret+= ip;
	ret+= '\n';
	
	ret+= "provider=";//obs:: n pode ter \n no provider
	ret+= provider;
	ret+= '\n';
	
	//adições para suporte de GPU
	ret+= "gpuMemoryTotal=";
	ret+= std::to_string(gpuMemoryTotal);
	ret+= '\n';
	
	ret+= "gpuMemoryMaxFrequency=";
	ret+= std::to_string(gpuMemoryMaxFrequency);
	ret+= '\n';
	
	ret+= "gpuMemoryBus=";
	ret+= std::to_string(gpuMemoryBus);
	ret+= '\n';
	
	ret+= "gpuMemoryBandwith=";
	ret+= std::to_string(gpuMemoryBandwith);
	ret+= '\n';
	
	ret+= "gpuFloatingPointPerf=";
	ret+= std::to_string(gpuFloatingPointPerf);
	ret+= '\n';
	
	ret+= "gpuMaxFrequency=";
	ret+= std::to_string(gpuMaxFrequency);
	ret+= '\n';
	
	return ret;
}


PluginInfo::PluginInfo(std::string const &str){
/*
	char tempID[TEMP_BUFFER_SIZE], tempInstanceName[TEMP_BUFFER_SIZE], 
	int res= sscanf(
				str.c_str(),
				"PLUGIN_INFO id=%s instanceName=%s privateCloud=%d host=%[^:]:%d upTime=%ld latency=%lf costPerGiga=%lf timestamp=%ld numCores=%d numNodes=%d numOccupied=%d ranking=%lf fsSize=%f memoryTotal=%lf memoryFree=%lf currentFrequencyCore=%lf costPerHour=%lf bandwith=%lf ip=%s provider=%s gpuMemoryTotal=%lf gpuMemoryMaxFrequency=%lf gpuMemoryBus=%lf gpuMemoryBandwith=%lf gpuFloatingPointPerf%lf gpuMaxFrequency=%lf"
				
				);
*/
	int size= str.length();
	char *temp= (char*)operator new[] (size+1);
	memcpy(temp, str.c_str(), size);
	temp[size]= '\0';
	
	char delimiter[2];
	delimiter[0]= '\n';
	delimiter[1]= '\0';
	char* aux= strtok(temp, delimiter);
	ASSERT2(0 == memcmp("PLUGIN_INFO" , aux, STRLEN("PLUGIN_INFO") )
							|| 0 == memcmp("value=PLUGIN_INFO" , aux, STRLEN("value=PLUGIN_INFO") ), "expecting PLUGIN_INFO, got " << temp );
	char *token, buffer[TEMP_BUFFER_SIZE+1];
	buffer[TEMP_BUFFER_SIZE]= '\0';
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "id=%[^\n]", buffer) );
	id= buffer;
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "instanceName=%[^\n]", buffer) );
	instanceName= buffer;
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "privateCloud=%d", &privateCloud) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(2 == sscanf(token, "host=%[^:]:%d", buffer, &(host.port) ) );
	host.adderss= buffer;
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "upTime=%ld", &upTime) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "latency=%lf", &latency) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "costPerGiga=%lf", &costPerGiga) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "timestamp=%ld", &timestamp) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "numCores=%d", &numCores) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "numNodes=%d", &numNodes) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "numOccupied=%d", &numOccupied) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "ranking=%lf", &ranking) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "fsSize=%f", &fsSize) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "memoryTotal=%lf", &memoryTotal) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "memoryFree=%lf", &memoryFree) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "currentFrequencyCore=%lf", &currentFrequencyCore) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "costPerHour=%lf", &costPerHour) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	bandwith= NAN;
//	ASSERT(1 == sscanf(token, "bandwith=%lf", &bandwith) );
	if(0 == sscanf(token, "bandwith=%lf", &bandwith)){
		if(0 == memcmp("bandwith=null", token, STRLEN("bandwith=null") ) ){
			bandwith= 0;
		}
	}
	if(NAN == bandwith){
		Error("Could not read bandwith, found it: " << token);
	}
	REPORT_DEBUG2(1, "\ttoken= " <<token<< "\tbandwith = "<< bandwith)
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "ip=%[^\n]", buffer) );
	ip= buffer;
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "provider=%[^\n]", buffer) );
	provider= buffer;
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "gpuMemoryTotal=%lf", &gpuMemoryTotal) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "gpuMemoryMaxFrequency=%lf", &gpuMemoryMaxFrequency) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "gpuMemoryBus=%lf", &gpuMemoryBus) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "gpuMemoryBandwith=%lf", &gpuMemoryBandwith) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "gpuFloatingPointPerf=%lf", &gpuFloatingPointPerf) );
	
	token= strtok(NULL, delimiter);
	REPORT_DEBUG("\ttoken= " << token << "\n");
	ASSERT(1 == sscanf(token, "gpuMaxFrequency=%lf", &gpuMaxFrequency) );
	
	operator delete[](temp);
	
}

bool PluginInfo::operator==(PluginInfo const &other){
	bool ret= true;
	
	COMPARE(id);
	ret= ret && (id==other.id);
	
	COMPARE(instanceName);
	ret= ret && (instanceName==other.instanceName);
	
	COMPARE(privateCloud);
	ret= ret && (privateCloud==other.privateCloud);
	
	BASIC_COMPARE(host);
	ret= ret && (host==other.host);
	
	COMPARE(upTime);
	ret= ret && (upTime==other.upTime);
	
	COMPARE(latency);
	ret= ret && (latency==other.latency);
	
	COMPARE(costPerGiga);
	ret= ret && (costPerGiga==other.costPerGiga);
	
	COMPARE(timestamp);
	ret= ret && (timestamp==other.timestamp);
	
	COMPARE(numCores);
	ret= ret && (numCores==other.numCores);
	
	COMPARE(ranking);
	ret= ret && (ranking==other.ranking);
	
	COMPARE(fsSize);
	ret= ret && (fsSize==other.fsSize);
	
	COMPARE(memoryTotal);
	ret= ret && (memoryTotal==other.memoryTotal);
	
	COMPARE(memoryFree);
	ret= ret && (memoryFree==other.memoryFree);
	
	COMPARE(currentFrequencyCore);
	ret= ret && (currentFrequencyCore==other.currentFrequencyCore);
	
	COMPARE(costPerHour);
	ret= ret && (costPerHour==other.costPerHour);
	
	COMPARE(bandwith);
	ret= ret && (bandwith==other.bandwith);
	
	COMPARE(ip);
	ret= ret && (ip==other.ip);
	
	COMPARE(provider);
	ret= ret && (provider==other.provider);
	
	COMPARE(gpuMemoryTotal);
	ret= ret && (gpuMemoryTotal==other.gpuMemoryTotal);
	
	COMPARE(gpuMemoryMaxFrequency);
	ret= ret && (gpuMemoryMaxFrequency==other.gpuMemoryMaxFrequency);
	
	COMPARE(gpuMemoryBus);
	ret= ret && (gpuMemoryBus==other.gpuMemoryBus);
	
	COMPARE(gpuMemoryBandwith);
	ret= ret && (gpuMemoryBandwith==other.gpuMemoryBandwith);
	
	COMPARE(gpuFloatingPointPerf);
	ret= ret && (gpuFloatingPointPerf==other.gpuFloatingPointPerf);
	
	COMPARE(gpuMaxFrequency);
	ret= ret && (gpuMaxFrequency==other.gpuMaxFrequency);
	
	return ret;
}

bool PluginInfo::TestSerialization(void){
	PluginInfo a;
	a.id= "idTeste";
	a.instanceName= "testInstanceName";
	a.privateCloud= 14;
	a.host= Host("1.2.3.4", 31415);
	a.upTime= 80040234823;
	a.latency= 123.56456;
	a.costPerGiga= 567.989;
	a.timestamp= 123;
	a.numCores= 2;
	a.numNodes=3;
	a.numOccupied=4;
	a.ranking= 2234;
	a.fsSize= 45456.234;
	a.memoryTotal= 765.4;
	a.memoryFree= 654.3;
	a.currentFrequencyCore= 12345.6;
	a.costPerHour=9.87;
	a.bandwith= 9.76;
	a.ip= "1.2.3.4";
	a.provider= "eu.mesmo";
	a.gpuMemoryTotal= 123.123;
	a.gpuMemoryMaxFrequency= 234.234;
	a.gpuMemoryBus= 345.345;
	a.gpuMemoryBandwith= 456.456;
	a.gpuFloatingPointPerf= 567.567;
	a.gpuMaxFrequency= 678.678;
	
	std::string serializated= a.Serialize();
	printf("\n------\n%s\n------\n", serializated.c_str());
	PluginInfo b(serializated);
	return a==b;
}


